File: sdk/node/machine.ts
/** The public `Machine` — an isolated microVM, local (embedded) or cloud.
 *
 *  The backend is chosen by `ConnectOptions`:
 *    - default / `{ target: 'local' }` → embedded engine, no server.
 *    - `{ target: 'cloud', apiKey }` (or `SMOL_CLOUD_TOKEN`) → smolfleet cloud.
 *
 *  Both go through the same `Transport`, so calling code is identical; features
 *  a given backend lacks throw `NotSupportedError`.
 */

import { randomUUID } from "node:crypto";
import { ExecutionError } from "./errors";
import {
  makeTransport,
  connectTransport,
  restoreCheckpointTransport,
  type RawExec,
  type Transport,
} from "./transport";
import type {
  AssignOptions,
  BranchBatchOptions,
  BranchOptions,
  ConnectOptions,
  ExecEvent,
  ExecOptions,
  ExecResult,
  ForkBatchOptions,
  ForkOptions,
  ImageInfo,
  MachineConfig,
  MachineUsageReport,
  PortableCheckpointInfo,
  PortEndpoint,
  PortSpec,
  WaitReadyOptions,
} from "./types";

function makeExecResult(r: RawExec): ExecResult {
  const success = r.exitCode === 0;
  return {
    exitCode: r.exitCode,
    stdout: r.stdout,
    stderr: r.stderr,
    // Cloud-only truncation flags (output capped at 1 MiB); the local engine
    // streams unbounded, so absent means false.
    stdoutTruncated: r.stdoutTruncated ?? false,
    stderrTruncated: r.stderrTruncated ?? false,
    // Byte-exact output. The cloud transport decodes it from base64; the local
    // transport has no bytes, so fall back to the UTF-8 encoding of the text.
    stdoutBytes: r.stdoutBytes ?? new TextEncoder().encode(r.stdout),
    stderrBytes: r.stderrBytes ?? new TextEncoder().encode(r.stderr),
    success,
    output: r.stdout + r.stderr,
    assertSuccess() {
      if (!success) throw new ExecutionError(r.exitCode, r.stdout, r.stderr);
    },
  };
}

export class Machine {
  private constructor(private readonly transport: Transport) {}

  /**
   * Create and start a machine. On the cloud target, this does not return until
   * the machine is ready for work: the guest agent is reachable and any
   * published port is accepting connections. A cloud state of `"started"`
   * alone means only that the VM process launched.
   *
   * @param config  machine configuration (a name is generated if omitted; `image`
   *                is the base image — required for cloud, optional for local)
   * @param conn    backend selection (local embedded by default)
   */
  static async create(
    config: MachineConfig = {},
    conn: ConnectOptions = {},
  ): Promise<Machine> {
    // The native/cloud transports still use the compatibility `forkable`
    // field. Normalize every accepted public spelling at this boundary.
    const branchable = config.branchable ?? config.forkable ?? config.checkpoint;
    if (branchable !== undefined) config = { ...config, forkable: branchable };
    return new Machine(await makeTransport(config, conn));
  }

  /**
   * Attach to an EXISTING machine without creating a new one — to drive a
   * machine made elsewhere (another process, the console, the REST API).
   *  - local (default): re-opens a persisted machine by NAME, starting it if
   *    stopped — pairs with `Machine.create({ name, … }, …)` + `persistent`.
   *  - cloud: looks up the machine by id; throws if it doesn't exist.
   *    This does not wait for readiness; call `waitUntilReady()` before `exec`,
   *    using a connect endpoint, or expecting the workload to respond.
   *
   * @param id    local machine name, or cloud machine id (`mach-…`)
   * @param conn  backend selection (local by default; cloud via `{ target: 'cloud', apiKey }` or `SMOL_CLOUD_TOKEN`)
   */
  static async connect(
    id: string,
    conn: ConnectOptions = {},
  ): Promise<Machine> {
    return new Machine(await connectTransport(id, conn));
  }

  /** Restore a local `.smolcheckpoint` path or durable cloud checkpoint id into
   * a ready, immediately forkable machine. */
  static async restoreCheckpoint(
    checkpointId: string,
    name: string,
    conn?: ConnectOptions,
  ): Promise<Machine> {
    return new Machine(await restoreCheckpointTransport(checkpointId, name, conn));
  }

  /** Restore a local `.smolcheckpoint` path or durable cloud checkpoint id.
   * Alias of {@link restoreCheckpoint}. */
  static restore(
    checkpointId: string,
    name: string,
    conn?: ConnectOptions,
  ): Promise<Machine> {
    return Machine.restoreCheckpoint(checkpointId, name, conn);
  }

  /** The machine's name / identifier. */
  get name(): string {
    return this.transport.name;
  }

  /** The machine's id — the cloud `mach-…` id (or the name on local). Handy right
   *  after `create()`/`fork()` so you don't have to list the tenant to find the
   *  machine you just made. */
  get id(): string {
    return this.transport.machineId;
  }

  /** Current lifecycle state. On cloud, `"started"` means only that the VM
   *  process launched; it does not mean the guest or workload is ready. Use
   *  `ready()` or `waitUntilReady()` before doing work. */
  state(): Promise<string> {
    return this.transport.state();
  }

  /** Whether the machine is READY to do work. `state()` becoming "started"
   *  means only that the VM process launched — the guest is still booting and
   *  is NOT yet usable. `ready` becomes true once the in-VM agent is reachable
   *  (an `exec`/`connect` will succeed) and any published port accepts
   *  connections. Gate on this, not `state`, before driving the machine.
   *  (cloud; the local target reports ready once running.) */
  ready(): Promise<boolean> {
    return this.transport.ready();
  }

  /** When the machine first became ready (RFC3339), or `null` if not yet ready. */
  readyAt(): Promise<string | null> {
    return this.transport.readyAt();
  }

  /** Block until the machine is `ready` (or throw on a failed/stopped state or
   *  timeout). `create()` already waits for readiness, so this is for machines
   *  attached via `Machine.connect(...)`, or to re-assert readiness before use.
   *  Poll ceiling and interval are configurable (defaults: 120s / 1s). */
  waitUntilReady(opts?: WaitReadyOptions): Promise<void> {
    return this.transport.waitUntilReady(opts);
  }

  /** An endpoint (URL + any required headers) for a PUBLISHED guest port.
   *  Local machines resolve to their current localhost host-port mapping;
   *  cloud machines use the authenticated connect bridge. */
  endpoint(port: number, path?: string): PortEndpoint {
    return this.transport.endpoint(port, path);
  }

  /** Convenience HTTP request to a published guest port. Returns the raw
   *  `fetch` `Response` on local and cloud targets.
   *
   *  @param port  a published GUEST port the in-VM service listens on
   *  @param path  optional path on that service (e.g. `"healthz"`)
   *  @param init  standard `fetch` init; its headers merge over the auth header */
  fetch(port: number, path?: string, init?: RequestInit): Promise<Response> {
    const e = this.transport.endpoint(port, path);
    return fetch(e.httpUrl, {
      ...init,
      headers: {
        ...e.headers,
        ...((init?.headers as Record<string, string> | undefined) ?? {}),
      },
    });
  }

  /** Public ingress URL for the machine's first published port (cloud).
   *  `null` until the machine is started with an allocated host port, for
   *  machines with no published port, or on the local target (no public
   *  ingress). Reach the deployed app over HTTPS at the returned URL. */
  url(): Promise<string | null> {
    return this.transport.url();
  }

  /** Execute a command directly in the machine. */
  async exec(command: string[], opts?: ExecOptions): Promise<ExecResult> {
    return makeExecResult(await this.transport.exec(command, opts));
  }

  /** Pull an image (if needed) and run a command in a container of it. (local) */
  async run(
    image: string,
    command: string[],
    opts?: ExecOptions,
  ): Promise<ExecResult> {
    return makeExecResult(await this.transport.run(image, command, opts));
  }

  /** Execute a command, yielding stdout/stderr/exit events. (local) */
  execStream(command: string[], opts?: ExecOptions): AsyncGenerator<ExecEvent> {
    return this.transport.execStream(command, opts);
  }

  /** Read a file from the machine. */
  readFile(path: string): Promise<Buffer> {
    return this.transport.readFile(path);
  }

  /** Write a file into the machine. */
  writeFile(
    path: string,
    data: string | Uint8Array,
    mode?: number,
  ): Promise<void> {
    const buf =
      typeof data === "string" ? Buffer.from(data) : Buffer.from(data);
    return this.transport.writeFile(path, buf, mode);
  }

  /** Pull an OCI image into the machine's storage. (local) */
  pullImage(image: string): Promise<ImageInfo> {
    return this.transport.pullImage(image);
  }

  /** List cached OCI images. (local) */
  listImages(): Promise<ImageInfo[]> {
    return this.transport.listImages();
  }

  /** Stop the machine. */
  stop(): Promise<void> {
    return this.transport.stop();
  }

  /** Start (resume) a stopped machine, awaiting until its agent is ready. The
   *  counterpart to `stop()` — disk state is preserved across a stop/start, so
   *  this is a cheap pause/resume. */
  start(): Promise<void> {
    return this.transport.start();
  }

  /** Stop the machine and delete its storage. On the cloud target, pass
   *  `{ includeUsage: true }` to also get the fully settled usage + cost back
   *  in one call — the control plane takes a final metering sample before
   *  teardown, so nothing accrues after the returned report. */
  delete(): Promise<void>;
  delete(opts: { includeUsage: true }): Promise<MachineUsageReport>;
  delete(opts?: {
    includeUsage?: boolean;
  }): Promise<MachineUsageReport | void> {
    if (opts?.includeUsage) return this.transport.deleteWithUsage();
    return this.transport.delete();
  }

  /** Metered usage + cost for this machine (cloud target). Usage records
   *  survive deletion for 30 days. CPU/memory accrue via a rollup that samples
   *  every few minutes, so a mid-life read is a lower bound; the report is
   *  final once the machine stops or is deleted. */
  usage(): Promise<MachineUsageReport> {
    return this.transport.usage();
  }

  /** Capture this running checkpointable machine. Local capture requires an
   * output `.smolcheckpoint` path; cloud capture stores the artifact durably. */
  checkpoint(output?: string): Promise<PortableCheckpointInfo> {
    return this.transport.checkpoint(output);
  }

  /** List durable portable checkpoints captured from this machine. */
  checkpoints(): Promise<PortableCheckpointInfo[]> {
    return this.transport.checkpoints();
  }

  /** Branch an independent child from this running source. The child inherits
   *  warm RAM and disk through copy-on-write while the source remains available
   *  for more branches. Create the source with `{ branchable: true }`.
   *
   *  @param name name for the child machine.
   *  @param options optional pinned ports and `branchable: true` when the child
   *                 must itself become a branch source. */
  branch(name: string, options?: PortSpec[] | BranchOptions): Promise<Machine> {
    return this.transport.fork(name, options).then((transport) => new Machine(transport));
  }

  /** @deprecated Use {@link branch}. */
  fork(name: string, options?: PortSpec[] | ForkOptions): Promise<Machine> {
    return this.branch(name, options);
  }

  /** Branch this source into MANY independent children in one call. On the
   *  cloud target the batch is transactional, so callers receive all children
   *  or none. Create the source with `{ branchable: true }`.
   *
   *  @param opts  batch size (`count` or `names`), optional `namePrefix`/`ports`
   *  @returns the children, in request order
   *
   *  ```ts
   *  const children = await source.branchBatch({ count: 32, namePrefix: "rollout" });
   *  await Promise.all(children.map((child) => child.exec(["python", "rollout.py"])));
   *  ``` */
  branchBatch(opts: BranchBatchOptions): Promise<Machine[]> {
    return this.transport
      .forkBatch(opts)
      .then((children) => children.map((transport) => new Machine(transport)));
  }

  /** @deprecated Use {@link branchBatch}. */
  async forkBatch(opts: ForkBatchOptions): Promise<Machine[]> {
    return this.branchBatch(opts);
  }

  /** Assign an RL episode: fork + provision a clone of this forkable golden under
   *  an idempotent lease, returned only once fully installed — with lifecycle
   *  control (`Episode.heartbeat()` / `Episode.complete()`). The scheduler's "give
   *  me one clean worker for this task, exactly once" call. **Cloud target only.**
   *
   *  Idempotent on `opts.leaseId`: a retried assign with the same key returns the
   *  SAME episode instead of a second fork. `task` is staged into the clone at
   *  `/run/smol-task.json`, `secrets` at `/run/smol-secrets.json`, and `files` at
   *  their paths — all before the episode is handed back. Always `complete()` the
   *  episode (e.g. in a `finally`) so its clone is reclaimed.
   *
   *  ```ts
   *  const ep = await golden.assign({ leaseId: "task-42", task: { seed: 7 } });
   *  try {
   *    await ep.exec(["python", "rollout.py"]);
   *  } finally {
   *    await ep.complete("done");
   *  }
   *  ``` */
  async assign(opts: AssignOptions): Promise<Episode> {
    const { transport, leaseId, ownerToken } = await this.transport.assign(opts);
    return new Episode(new Machine(transport), leaseId, ownerToken, this.transport);
  }

  /** Score this machine's state WITHOUT mutating it: fork an ephemeral clone,
   *  run the grader command in the clone, then destroy it. The result (exit code
   *  / stdout) is your reward signal; this machine is untouched. This is the RL
   *  "reward judging without side effects" primitive — grade the end of a rollout,
   *  or run a verifier — as one call, with the throwaway clone always cleaned up
   *  even if the grader errors. Requires this machine to be `forkable` (like
   *  `fork()`); a clone name is generated unless you pass one.
   *
   *  @param command  the grader/verifier argv to run in the clone
   *  @param opts      exec options (env, cwd, timeout, …)
   *  @param name      optional clone name (default: a generated unique name)
   *  @returns the grader's `ExecResult` (use `.success` for pass/fail, `.stdout` for a score) */
  async rewardFork(
    command: string[],
    opts?: ExecOptions,
    name?: string,
  ): Promise<ExecResult> {
    const cloneName = name ?? `${this.name}-rf-${randomUUID().slice(0, 8)}`;
    const clone = await this.fork(cloneName);
    try {
      return await clone.exec(command, opts);
    } finally {
      // Best-effort teardown — never mask the grader's result/error.
      await clone.delete().catch(() => {});
    }
  }
}

/** A leased RL episode: a freshly-provisioned clone plus its lease. Created by
 *  `Machine.assign()`. Always `complete()` it (e.g. in a `finally`) so the clone
 *  is reclaimed — the RL "one clean worker per task, guaranteed reclaimed"
 *  pattern. Cloud target only. */
export class Episode {
  readonly #machine: Machine;
  readonly leaseId: string;
  readonly #ownerToken: string;
  readonly #leaseTransport: Transport;
  #completed = false;

  constructor(
    machine: Machine,
    leaseId: string,
    ownerToken: string,
    leaseTransport: Transport,
  ) {
    this.#machine = machine;
    this.leaseId = leaseId;
    this.#ownerToken = ownerToken;
    this.#leaseTransport = leaseTransport;
  }

  /** The episode's provisioned clone. */
  get machine(): Machine {
    return this.#machine;
  }

  /** Run a command in the episode's machine (shortcut for `.machine.exec`). */
  async exec(command: string[], opts?: ExecOptions): Promise<ExecResult> {
    return this.#machine.exec(command, opts);
  }

  /** Keep the lease alive. Call within the `heartbeatSecs` cadence you requested,
   *  or the episode is reclaimed as `heartbeat_lost`. */
  async heartbeat(): Promise<void> {
    await this.#leaseTransport.heartbeatLease(this.leaseId, this.#ownerToken);
  }

  /** Finish the episode with a typed termination reason (e.g. `done`,
   *  `agent_failed`, `infra_failed`, `cancelled`) and tear its clone down.
   *  Optionally record a `score` (per-task reward / pass-rate) and an arbitrary
   *  JSON `result` — the "did it improve?" number, read back via `status()`.
   *  Idempotent — a second call is a no-op. */
  async complete(
    reason = "done",
    opts?: { score?: number; result?: unknown },
  ): Promise<void> {
    if (this.#completed) return;
    await this.#leaseTransport.completeLease(
      this.leaseId,
      this.#ownerToken,
      reason,
      opts?.score,
      opts?.result,
    );
    this.#completed = true;
  }

  /** Read this lease's current status and, once completed, its outcome (`state`,
   *  `reason`, `score`, `result`) — how a trainer collects the per-task score. */
  async status(): Promise<Record<string, unknown>> {
    return this.#leaseTransport.getLease(this.leaseId);
  }
}

File: sdk/rust/src/machine.rs
//! The machine handle: create, start, run things, fork, checkpoint, tear down.

use std::path::{Path, PathBuf};
use std::sync::Arc;

use smolvm::embedded::{runtime, EmbeddedRuntime};

use crate::config::{MachineBuilder, MachineConfig, Port};
use crate::error::Result;
use crate::exec::{ExecOptions, ExecResult, ExecStream};

/// Where a machine is in its lifecycle.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[non_exhaustive]
pub enum MachineState {
    /// Created or stopped; no VM process.
    Stopped,
    /// Booting, or waiting for the guest agent.
    Starting,
    /// Booted, agent connected.
    Running,
    /// Shutting down.
    Stopping,
}

impl MachineState {
    fn parse(state: &str) -> Self {
        match state {
            "starting" => Self::Starting,
            "running" => Self::Running,
            "stopping" => Self::Stopping,
            _ => Self::Stopped,
        }
    }

    /// The engine's own name for this state.
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Stopped => "stopped",
            Self::Starting => "starting",
            Self::Running => "running",
            Self::Stopping => "stopping",
        }
    }
}

impl std::fmt::Display for MachineState {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.as_str())
    }
}

/// An OCI image cached in a machine's storage.
#[derive(Debug, Clone)]
pub struct ImageInfo {
    /// Image reference, as pulled.
    pub reference: String,
    /// Content digest.
    pub digest: String,
    /// Size on disk in bytes.
    pub size: u64,
    /// Image architecture.
    pub architecture: String,
    /// Image OS.
    pub os: String,
}

impl From<smolvm_protocol::ImageInfo> for ImageInfo {
    fn from(info: smolvm_protocol::ImageInfo) -> Self {
        Self {
            reference: info.reference,
            digest: info.digest,
            size: info.size,
            architecture: info.architecture,
            os: info.os,
        }
    }
}

/// Where a checkpoint writes, and what it may reuse.
#[derive(Debug, Clone, Default)]
pub struct CheckpointOptions {
    /// A content-addressed store to write chunks into. Reusing one store across
    /// captures of the same machine makes each later capture incremental: only
    /// changed chunks are written, and the rest are hard-linked. This is the
    /// difference between a full save and a few seconds.
    pub store_dir: Option<PathBuf>,
    /// Scratch directory for capture. Defaults to the store's own staging area.
    pub staging_dir: Option<PathBuf>,
    /// Cap on the prepared-chunk cache, in bytes.
    pub prepared_cache_budget_bytes: Option<u64>,
}

impl CheckpointOptions {
    /// Options with nothing set.
    pub fn new() -> Self {
        Self::default()
    }

    /// Write chunks into this store, reusing anything already there.
    pub fn store_dir(mut self, dir: impl Into<PathBuf>) -> Self {
        self.store_dir = Some(dir.into());
        self
    }

    /// Use this scratch directory during capture.
    pub fn staging_dir(mut self, dir: impl Into<PathBuf>) -> Self {
        self.staging_dir = Some(dir.into());
        self
    }

    /// Cap the prepared-chunk cache.
    pub fn prepared_cache_budget_bytes(mut self, bytes: u64) -> Self {
        self.prepared_cache_budget_bytes = Some(bytes);
        self
    }
}

/// What a capture cost and produced.
#[derive(Debug, Clone)]
pub struct CheckpointResult {
    /// Size of the artifact in bytes.
    pub size_bytes: u64,
    /// Bytes served from the store rather than written again. Zero without a
    /// store, and most of the artifact on a warm incremental capture.
    pub reused_bytes: u64,
    /// How long the source machine was paused. This is the only part the
    /// workload notices.
    pub source_pause: std::time::Duration,
    /// Wall-clock time for the whole capture.
    pub elapsed: std::time::Duration,
}

impl From<smolvm::portable_checkpoint::CaptureResult> for CheckpointResult {
    fn from(result: smolvm::portable_checkpoint::CaptureResult) -> Self {
        Self {
            size_bytes: result.size_bytes,
            reused_bytes: result.reused_bytes,
            source_pause: result.source_pause,
            elapsed: result.elapsed,
        }
    }
}

/// How a branch is made.
#[derive(Debug, Clone, Default)]
pub struct BranchOptions {
    /// Inbound port forwards for the branch. A branch does not inherit its
    /// source's forwards, because two machines cannot listen on one host port.
    pub ports: Vec<Port>,
    /// Make the branch itself checkpointable, at some cost to branch time.
    pub checkpointable: bool,
    /// How many branches to boot at once in [`Machine::branch_batch`].
    pub parallel: usize,
}

impl BranchOptions {
    /// Options with nothing set.
    pub fn new() -> Self {
        Self::default()
    }

    /// Forward a host port into the branch.
    pub fn port(mut self, port: Port) -> Self {
        self.ports.push(port);
        self
    }

    /// Make the branch checkpointable in turn.
    pub fn checkpointable(mut self, checkpointable: bool) -> Self {
        self.checkpointable = checkpointable;
        self
    }

    /// Boot this many branches at once during a batch branch.
    pub fn parallel(mut self, parallel: usize) -> Self {
        self.parallel = parallel;
        self
    }

    fn pinned(&self) -> Vec<(u16, u16)> {
        self.ports.iter().map(|p| (p.host, p.guest)).collect()
    }
}

/// How many branches boot at once when a batch branch does not say.
const DEFAULT_BRANCH_PARALLEL: usize = 8;

/// A handle to one machine.
///
/// The handle is just a name plus access to the process-wide runtime, so it is
/// cheap to clone and every handle for a given name talks to the same VM. It
/// owns nothing: dropping it does not stop or delete the machine.
#[derive(Debug, Clone)]
pub struct Machine {
    name: String,
}

fn rt() -> Result<Arc<EmbeddedRuntime>> {
    Ok(runtime()?)
}

impl Machine {
    /// Start describing a machine of this name.
    pub fn builder(name: impl Into<String>) -> MachineBuilder {
        MachineBuilder::new(name)
    }

    /// Create a machine from a config. The VM is not started yet.
    pub fn create(config: MachineConfig) -> Result<Self> {
        let name = config.name.clone();
        let request = config.into_request()?;
        rt()?.create_machine_with_workload(
            request.spec,
            request.env,
            request.workdir,
            request.user,
        )?;
        Ok(Self { name })
    }

    /// Attach to an existing machine, starting it if it is stopped.
    ///
    /// This is how a persisted machine is reopened from a new process.
    pub fn connect(name: impl Into<String>) -> Result<Self> {
        let name = name.into();
        rt()?.connect_or_start_machine(&name)?;
        Ok(Self { name })
    }

    /// Attach to an existing machine without starting it.
    ///
    /// Use this to inspect or delete a stopped machine, or to get a handle for
    /// one you know is already running.
    pub fn attach(name: impl Into<String>) -> Self {
        Self { name: name.into() }
    }

    /// Create a stopped machine from a portable checkpoint on disk.
    pub fn restore_checkpoint(name: impl Into<String>, artifact: impl AsRef<Path>) -> Result<Self> {
        let name = name.into();
        rt()?.restore_checkpoint_machine(&name, artifact.as_ref())?;
        Ok(Self { name })
    }

    /// The machine's name.
    pub fn name(&self) -> &str {
        &self.name
    }

    /// The VM process id, if it is running.
    pub fn pid(&self) -> Option<i32> {
        runtime().ok().and_then(|runtime| runtime.pid(&self.name))
    }

    /// Whether the VM process is alive.
    pub fn is_running(&self) -> bool {
        runtime()
            .map(|runtime| runtime.is_running(&self.name))
            .unwrap_or(false)
    }

    /// Where the machine is in its lifecycle.
    pub fn state(&self) -> MachineState {
        runtime()
            .map(|runtime| MachineState::parse(&runtime.state(&self.name)))
            .unwrap_or(MachineState::Stopped)
    }

    /// The host port forwarding to a published guest port.
    pub fn host_port(&self, guest_port: u16) -> Result<Option<u16>> {
        Ok(rt()?.host_port(&self.name, guest_port)?)
    }

    /// Every published guest port.
    pub fn guest_ports(&self) -> Result<Vec<u16>> {
        Ok(rt()?.guest_ports(&self.name)?)
    }

    /// Boot the VM and wait for the guest agent to answer.
    pub fn start(&self) -> Result<()> {
        Ok(rt()?.start_machine(&self.name)?)
    }

    /// Boot the VM as a branch source, with memfd-backed guest RAM and a
    /// control socket, so [`Machine::branch`] can clone it later.
    pub fn start_branchable(&self) -> Result<()> {
        Ok(rt()?.start_forkable_machine(&self.name)?)
    }

    /// Run a command in the guest and wait for it.
    pub fn exec<I, S>(&self, command: I) -> Result<ExecResult>
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        self.exec_with(command, ExecOptions::new())
    }

    /// Run a command in the guest with explicit environment, directory or
    /// timeout.
    pub fn exec_with<I, S>(&self, command: I, options: ExecOptions) -> Result<ExecResult>
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        let command: Vec<String> = command.into_iter().map(Into::into).collect();
        let (env, workdir, timeout) = options.split();
        let (exit_code, stdout, stderr) = rt()?.exec(&self.name, command, env, workdir, timeout)?;
        Ok(ExecResult {
            exit_code,
            stdout,
            stderr,
        })
    }

    /// Run a command in the guest and read its output as it arrives.
    pub fn exec_stream<I, S>(&self, command: I, options: ExecOptions) -> ExecStream
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        let command: Vec<String> = command.into_iter().map(Into::into).collect();
        let (env, workdir, timeout) = options.split();
        ExecStream::spawn(self.name.clone(), command, env, workdir, timeout)
    }

    /// Pull an OCI image if needed, run a command inside it on an overlay
    /// rootfs, then clean up. The equivalent of `smolvm run`.
    pub fn run<I, S>(&self, image: &str, command: I) -> Result<ExecResult>
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        self.run_with(image, command, ExecOptions::new())
    }

    /// [`Machine::run`] with explicit environment, directory or timeout.
    pub fn run_with<I, S>(
        &self,
        image: &str,
        command: I,
        options: ExecOptions,
    ) -> Result<ExecResult>
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        let command: Vec<String> = command.into_iter().map(Into::into).collect();
        let (env, workdir, timeout) = options.split();
        let (exit_code, stdout, stderr) =
            rt()?.run(&self.name, image, command, env, workdir, timeout)?;
        Ok(ExecResult {
            exit_code,
            stdout,
            stderr,
        })
    }

    /// Pull an OCI image into the machine's storage.
    pub fn pull_image(&self, image: &str) -> Result<ImageInfo> {
        Ok(rt()?.pull_image(&self.name, image)?.into())
    }

    /// Every OCI image cached in the machine's storage.
    pub fn list_images(&self) -> Result<Vec<ImageInfo>> {
        Ok(rt()?
            .list_images(&self.name)?
            .into_iter()
            .map(ImageInfo::from)
            .collect())
    }

    /// Write a file into the running guest.
    pub fn write_file(&self, path: &str, data: impl Into<Vec<u8>>) -> Result<()> {
        Ok(rt()?.write_file(&self.name, path, data.into(), None)?)
    }

    /// Write a file into the running guest with an explicit mode.
    pub fn write_file_with_mode(
        &self,
        path: &str,
        data: impl Into<Vec<u8>>,
        mode: u32,
    ) -> Result<()> {
        Ok(rt()?.write_file(&self.name, path, data.into(), Some(mode))?)
    }

    /// Read a file out of the running guest.
    pub fn read_file(&self, path: &str) -> Result<Vec<u8>> {
        Ok(rt()?.read_file(&self.name, path)?)
    }

    /// Capture this running machine to a portable checkpoint on disk.
    pub fn checkpoint(&self, output: impl AsRef<Path>) -> Result<CheckpointResult> {
        self.checkpoint_with(output, CheckpointOptions::new())
    }

    /// Capture this running machine, reusing a content-addressed store.
    ///
    /// Pointing every capture of a machine at the same store makes all but the
    /// first incremental.
    pub fn checkpoint_with(
        &self,
        output: impl AsRef<Path>,
        options: CheckpointOptions,
    ) -> Result<CheckpointResult> {
        let capture = smolvm::portable_checkpoint::CaptureOptions {
            rootfs_dir: Some(smolvm::agent::AgentManager::default_rootfs_path()?),
            store_dir: options.store_dir,
            staging_dir: options.staging_dir,
            prepared_cache_budget_bytes: options.prepared_cache_budget_bytes,
            ..Default::default()
        };
        Ok(rt()?
            .checkpoint_machine(&self.name, output.as_ref(), &capture)?
            .into())
    }

    /// Branch this running, branchable machine, sharing its live RAM and disks
    /// copy-on-write. Returns a handle to the running branch.
    pub fn branch(&self, name: impl Into<String>) -> Result<Machine> {
        self.branch_with(name, BranchOptions::new())
    }

    /// [`Machine::branch`] with port forwards, or a branch that is itself
    /// checkpointable.
    pub fn branch_with(&self, name: impl Into<String>, options: BranchOptions) -> Result<Machine> {
        let name = name.into();
        let pinned = options.pinned();
        let runtime = rt()?;
        if options.checkpointable {
            runtime.fork_checkpointable_machine(&self.name, &name, &pinned)?;
        } else {
            runtime.fork_machine(&self.name, &name, &pinned)?;
        }
        Ok(Machine { name })
    }

    /// Branch this machine many times from one retained checkpoint, booting the
    /// branches in bounded parallel waves.
    ///
    /// The call is transactional: if any branch fails, every branch made by
    /// this call is removed before the error returns.
    pub fn branch_batch<I, S>(&self, names: I, options: BranchOptions) -> Result<Vec<Machine>>
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        let names: Vec<String> = names.into_iter().map(Into::into).collect();
        let pinned = options.pinned();
        let parallel = if options.parallel == 0 {
            DEFAULT_BRANCH_PARALLEL
        } else {
            options.parallel
        };
        rt()?.fork_machines(&self.name, &names, &pinned, parallel)?;
        Ok(names.into_iter().map(|name| Machine { name }).collect())
    }

    /// Copy guest-local staged mounts back to their host sources.
    pub fn sync(&self) -> Result<()> {
        Ok(rt()?.sync_machine(&self.name)?)
    }

    /// Shut the VM down gracefully, keeping its disks.
    pub fn stop(&self) -> Result<()> {
        Ok(rt()?.stop_machine(&self.name)?)
    }

    /// Stop the VM and remove its disks and config. Not reversible.
    pub fn delete(&self) -> Result<()> {
        Ok(rt()?.delete_machine(&self.name)?)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn an_unknown_state_string_reads_as_stopped() {
        assert_eq!(MachineState::parse("running"), MachineState::Running);
        assert_eq!(MachineState::parse("starting"), MachineState::Starting);
        assert_eq!(MachineState::parse("stopping"), MachineState::Stopping);
        assert_eq!(MachineState::parse("who knows"), MachineState::Stopped);
    }

    #[test]
    fn every_state_round_trips_through_its_engine_name() {
        for state in [
            MachineState::Stopped,
            MachineState::Starting,
            MachineState::Running,
            MachineState::Stopping,
        ] {
            assert_eq!(MachineState::parse(state.as_str()), state);
        }
    }

    #[test]
    fn a_batch_branch_without_a_width_still_boots_in_waves() {
        assert_eq!(BranchOptions::new().parallel, 0);
        let options = BranchOptions::new().parallel(4);
        assert_eq!(options.parallel, 4);
        assert_eq!(
            BranchOptions::new().port(Port::new(9000, 80)).pinned(),
            vec![(9000, 80)]
        );
    }

    #[test]
    fn a_handle_never_owns_the_machine_it_names() {
        let machine = Machine::attach("detached");
        assert_eq!(machine.name(), "detached");
        assert_eq!(machine.clone().name(), "detached");
    }
}
